core/src/main/scala/cats/Iso.scala

```diff
@@ -0,0 +1,119 @@
+package cats
+
+import cats.Iso.HasIso
+import cats.arrow.{ArrowChoice, Category, Strong}
+import cats.evidence.{=:!=, =~~=}
+import cats.implicits._
+
+trait Iso[->[_, _], A, B] { self =>
+  def cat: Category[->]
+  def to: A -> B
+  def from: B -> A
+
+  final def apply(a: A)(implicit ev: -> =~~= Function1): B = ev(to)(a)
+
+  final def andThen[C](bc: Iso[->, B, C]): Iso[->, A, C] =
+    Iso.unsafe(cat.andThen(self.to, bc.to), cat.andThen(bc.from, self.from))(cat)
+
+  final def compose[Z](za: Iso[->, Z, A]): Iso[->, Z, B] = za.andThen(self)
+
+  /** Flips the isomorphism from A <-> B to B <-> A grace to it's reflexivity property */
+  def flip: Iso[->, B, A] = new Iso[->, B, A] {
+    val (cat, to, from) = (self.cat, self.from, self.to)
+    override val flip = self
+  }
+
+  /** If A <-> B then having a function B -> B we can obtain A -> A */
+  def teleport(fn: B -> B): A -> A = cat.compose(from, cat.compose(fn, to))
+}
+
+object Iso {
+  type <=>[A, B] = Iso[* => *, A, B]
+
+  /** Create an isomorphism from (A -> B, B -> A) as long as you promise to abide the isomorphism laws */
+  def unsafe[->[_, _], A, B](ab: A -> B, ba: B -> A)(implicit C: Category[->]): Iso[->, A, B] =
+    new Iso[->, A, B] { val cat = C; val (to, from) = (ab, ba) }
+
+  /** Isomorphism is reflexive: for any A we have A <-> A */
+  def refl[A]: Iso[* => *, A, A] = reflAny.asInstanceOf
+  private[this] val reflAny = new Iso[* => *, Any, Any] {
+    val (cat, to, from) = (Category[* => *], (a: Any) => a, (a: Any) => a)
+  }
+
+  def refl[->[_, _], A](implicit C: Category[->]): Iso[->, A, A] =
+    new Iso[->, A, A] { val (cat, to, from) = (C, C.id[A], C.id[A]) }
+
+  implicit final class IsoOps[->[_, _], A, B](val self: Iso[->, A, B]) extends AnyVal {
```

Comment:
Why not just put these methods on `Iso`?

Reply:
good point